Add type declaration

FranklinWhale opened the pull request:
Files for #126

FranklinWhale commented:
@okcoker Would you re-run the build again? I have no idea why it timed out, as I am able to build the UMD supported version on my local PC successfully. Thanks.

FranklinWhale commented:
Great. Successfully rebuilt with no file change by closing then re-opening this PR.

okcoker commented:
Hey @FranklinWhale, few things:

- Files should have newlines at the end of them as noted in the .editorconfig
- I don't think there's any reason for the css, example/js/taggle.min.js or example/js/scripts.js to change
- Why was @types/jquery and @types/jquery-ui added? This library doesnt depend on them.

Other than that, the types look good to me.

FranklinWhale commented:
@okcoker: Replies below:

1. I have added newline at the end of the new files.
2. The css files and taggle.min.js are changed by Grunt. Perhaps I should find a way to revert them. The changes in scripts.js are however necessary to make it compilable by TypeScript compiler. To keep taggle.d.ts in sync, I have added a test script to run the TypeScript compiler to compile scripts.js and taggle.d.ts together without any output. I expect that when some interfaces or properties in taggle.js are changed, scripts.js will be updated. The change details are as follows:
  - Replace `window.exampleN =` with `var exampleN =` because it is not allowed to set variable to `Window` directly
  - Add a `@ts-ignore` directive above `e.which` to workaround an error caused by the type declaration of jQueryUI
3. @types/jquery and @types/jquery-ui are necessary for the TypeScript compiler to compile scripts.js because jQuery and jQueryUI are used there.

By the way, should I commit package-lock.json?

okcoker commented:
> Replace window.exampleN = with var exampleN = because it is not allowed to set variable to Window directly

> @types/jquery and @types/jquery-ui are necessary for the TypeScript compiler to compile scripts.js because jQuery and jQueryUI are used there.

Ok so the reason I put these on the window is for convenience for anyone playing with the library in their console. Would the script validation still pass if we removed the IIFE from around everything in that file? Based on your second comment though, can we just ignore scripts.js altogether? It's not part of the lib, just there as a part of the example. That would remove the need for the extra packages and changes to script.js.


> By the way, should I commit package-lock.json?

Meh, I don't think its necessary right now. I may add one on the next update or at least a yarn.lock file. I actually want to get rid of grunt as well and just use npm scripts but that's another story.

FranklinWhale commented:
> can we just ignore scripts.js altogether? It's not part of the lib, just there as a part of the example.

Um... I currently treat scripts.js as a test file of taggle.d.ts. It is possible to create a test file that does not use jQuery and jQueryUI, but when a public interface or property in taggle.js is changed, you may have to update the test file, in addition to taggle-test.js and script.js.

Is a new test file better?

okcoker commented:
This may be my lack of typescript knowledge but my confusion is why does taggle.d.ts have 

```js
"files": [
    "../example/js/scripts.js"
]
```

instead of 

```js
"files": [
    "../src/taggle.js"
]
```

If more examples are created with different autocomplete libraries like [typeahead](https://twitter.github.io/typeahead.js/) for example, it seems we'd need to add more type definitions right? It doesn't make sense to me that I would be adding types for other libraries in _this_ library. That seems like the equivalent of me adding tests for jQuery UI in this library. I was only expecting to add definitions added for taggle.js itself while everything in the example folder should be more of a ts-ignore/eslintignore/gitignore/npmignore situation. Does that make sense?

FranklinWhale commented:
@okcoker Done. Please review. Thanks.

FranklinWhale commented:
No problem. Users willing to try the type definition may install taggle via `npm install okcoker/taggle.js`.
